File: 046. 全排列.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        self.res = []

        def backtrack(path):
            if len(path) == len(nums):
                self.res.append(path[:])
                return
            for num in nums:
                if num in path:
                    continue
                path.append(num)
                backtrack(path)
                path.pop()

        backtrack([])
        return self.res

File: 046. 全排列.go
package main

func permute(nums []int) [][]int {
	res := make([][]int, 0)
	path := make([]int, 0)

	var backtrack func()
	backtrack = func() {
		if len(path) == len(nums) {
			temp := make([]int, len(path))
			copy(temp, path)
			res = append(res, temp)
		}

	outer:
		for i := range len(nums) {
			for j := range len(path) {
				if nums[i] == path[j] {
					continue outer
				}
			}

			path = append(path, nums[i])
			backtrack()
			path = path[:len(path)-1]
		}
	}

	backtrack()
	return res
}
